src/channel/listener.rs

```diff
@@ -687,9 +681,7 @@ impl State {
                     State::Left
                 }
                 Connectivity::Disconnected(disconnected) => match disconnected {
-                    Disconnected::ServerDisconnected | Disconnected::Disconnect => {
-                        State::WaitingForSocketToConnect { rejoin: None }
-                    }
+                    Disconnected::Disconnect => State::WaitingForSocketToConnect { rejoin: None },
```

Comment:
Instead of using nested match expressions, this might be more intuitive to match over (connectivity, self) top level because then this branch can be covered once by (Connectivity::Disconnected(Disconnected::Disconnect), _) => State::WaitingForSocketToConnect { rejoin: None }

Reply:
Yeah, when I first wrote this I debating which to be the inner and which to be the outer.  I'll try reworking it and see if it becomes more compact and less repetitive in the leaves.

Reply:
Rewriting the code, I realize there are places where I don't send errors to some of the `Sender`s for some of the states, so I'm going to fix those and in such case there isn't as much compression.

Reply:
Merging as in, so I can fix that bug in another issue. I'll look at refactoring and compressing again after that bug is fixed.